### ISSUE TITLE
Get current state inside the payload

### ISSUE THREAD

**Issue by samislam** [reduxjs/redux-toolkit#3122]
In react, if we want to make a calculation which is based on a property set inside the state, we can write the `setState` as follows:
```js
const [counter, setCounter] = useState(5)
// instead of this: 
// setCounter(counter + 1)
// you can do this:
setCounter((previousState)=> previousState + 1)
```
This is one of the fundamentals in react.

I am having a *complex* problem in my code, *let's say*, so I have no idea how to explain it in a simple question, but let's try to hide all that for now and see if it requires posting the details later 😁.

Is there a similar thing in redux toolkit where I can for instance do something as follows:
```jsx
import { setCounter } from 'features/posSlice'

const MyComponent = () => {
  const dispatch = useDispatch()
  dispatch(setCounter((previousState) => previousState + 1))
}
```

Now you might ask, *why you don't implement this inside the reducers?* so by doing something as follows:
```js
import { createSlice } from '@reduxjs/toolkit'

const posSlice = createSlice({
  name: 'pos',
  initialState: {
    counter: 5
  },
  reducers: {
    incrementBy = (state, { payload }) => {
       state.counter = state.counter + payload
    }
  }
})
```

I would say, **okay…** that's a good question, the reason is, not easy to explain because this is not the real example. 

Just for now, let's make it simple:

I would like to know if there's a method to get the current state *among all these async calls* TO get the **correct..current..state** in order to build upon it. from inside the controllers, not the models.

So, something like this inside the component, not the slice:

```jsx
import { setCounter } from 'features/posSlice'

const MyComponent = () => {
  const dispatch = useDispatch()
  dispatch(setCounter((previousState) => previousState + 1))
}
```

Is it possible? 

thanks 🌹

**Comment by phryneas** [comment]
I'm sorry, but the question is **really** "why you don't implement this inside the reducers".

Caclulation of new state should always happen in a reducer - you should only pass stuff you need for that calculation into the payload, never the new value. Please give the [Redux Style Guide](https://redux.js.org/style-guide/) a good read.

That said, you can do this in a thunk, although I would really recommend against it.

```js
dispatch((dispatch, getState) => {
  dispatch(setCounter(getState().yourCounterValue + 1))
})
```

**Comment by samislam** [comment]
> I'm sorry, but the question is **really** "why you don't implement this inside the reducers".
> 
> Caclulation of new state should always happen in a reducer - you should only pass stuff you need for that calculation into the payload, never the new value. Please give the [Redux Style Guide](https://redux.js.org/style-guide/) a good read.
> 
> That said, you can do this in a thunk, although I would really recommend against it.
> 
> ```js
> dispatch((dispatch, getState) => {
>   dispatch(setCounter(getState().yourCounterValue + 1))
> })
> ```

This doesn't work with Redux toolkit, right? 
just a minute, I'll give a brief overview of what I'm trying to achieve

**Comment by samislam** [comment]
Just a brief about the real problem, I am trying to make a React hook which handles the merging properties for Redux,

for instance, you'll have this slice:

```js
const createUserSlice = createSlice({
  name: "createUser",
  initialState:{
    name: "",
    firstName: "",
    lastName: "",
    // and 10 other properties
  }
})
```

So instead of having to define tens of reducers as follows:
```js
reducers: {
  setName = (state, { payload: name })=>{
    state.name = name
  },
  setFirstName = (state, { payload: firstName })=>{
    state.firstName = firstName
  },
  setLastName = (state, { payload: lastName })=>{
    state.lastName = lastName
  },
  // ... etc and 10 other reducers for the other properties
}
```
You can simply use the hook which works as:

```js
const [state, setState] = useReduxStates(createUserState, setUserState)
// ... then,
setState({ firstName: " Joyful Boy " })
setState({ lastName: " Boy ", firstName: "Joe" })
setState({ name: "klsndvkls " })
```
This method _merges_ the previous state with the payload.

**Comment by samislam** [comment]
> I'm sorry, but the question is **really** "why you don't implement this inside the reducers".
> 
> Caclulation of new state should always happen in a reducer - you should only pass stuff you need for that calculation into the payload, never the new value. Please give the [Redux Style Guide](https://redux.js.org/style-guide/) a good read.
> 
> That said, you can do this in a thunk, although I would really recommend against it.
> 
> ```js
> dispatch((dispatch, getState) => {
>   dispatch(setCounter(getState().yourCounterValue + 1))
> })
> ```

**This exactly solves the problem**, thank you, I wish it had type definitions though,
```
import { useDispatch } from 'react-redux'

// inside a component
const dispatch = useDispatch()
dispatch(()=>{
 // you won't get inteliSense here for the callback arguments
})
```

I'm not sure, though, why do you think it's not recommended

**Comment by phryneas** [comment]
So, `name` is just `firstName + lastName`? In that case, `name` should never be a part of your Redux store, but be a value you derive in a selector. Also, why do you think this could not be done in a reducer? The reducer would still have access to `firstName`, `lastName` and `name` (if you include that one, which you really should not).

> I'm not sure, though, why do you think it's not recommended

Because state should never be calculated outside of a reducer. An action should be an event a la "this happened", not "I handled everything, save this".

**Comment by phryneas** [comment]
> **This exactly solves the problem**, thank you, I wish it had type definitions though,

It will have type definitions if you use `useAppDispatch`, as shown in our [TypeScript Quick Start](https://redux-toolkit.js.org/tutorials/typescript#define-typed-hooks)

**Comment by phryneas** [comment]
Adding another thought: Logic like this should *exactly* be in a reducer and only there. What if you later call that setter in two different places, and after a while, the calculation logic in one changes, and you forget about that in the other place? This is business logic and should be handled in the business logic layer - a Redux reducer.

**Comment by samislam** [comment]
> import { useDispatch } from 'react-redux'
> 
> // inside a component
> const dispatch = useDispatch()
> dispatch(()=>{
>  // you won't get inteliSense here for the callback arguments
> })


Yes, The reason why I'm calculating it outside the reducer is because think of having too many other slices, 

`createProductSlice`, `createInvoiceSlice`, `createBranchSlice`, `createSubscriptionSlice` and all of those,

then having the merging objects logic inside every reducer would be a pain

**Comment by phryneas** [comment]
I cannot repeat how much I want you to read the style guide.
[Avoid putting form state in Redux](https://redux.js.org/style-guide/#avoid-putting-form-state-in-redux)

**Comment by samislam** [comment]
For instance, here's an example of how one of these reducers would look like:
```js
reducers: {
  setUserState = (state, { payload })=>{
  return { ...state, ...payload }
  }
}
```

```js
reducers: {
  setProductsState = (state, { payload })=>{
  return { ...state, ...payload }
  }
}
```

```js
reducers: {
  setInvoicesState = (state, { payload })=>{
  return { ...state, ...payload }
  }
}
```

```js
reducers: {
  setBranchesState = (state, { payload })=>{
  return { ...state, ...payload }
  }
}
```

**Comment by samislam** [comment]
> [Avoid Putting Form State In Redux](https://redux.js.org/style-guide/#avoid-putting-form-state-in-redux) 

Yes, this one is great, I really was in need of that advice 😂

**Comment by phryneas** [comment]
That's basically the absolute antipattern.

[Model actions as events, not setters](https://redux.js.org/style-guide/#model-actions-as-events-not-setters)

Please read the style guide and rethink what you are doing.

**Comment by markerikson** [comment]
Yeah. To echo what phryneas said:

You're trying to solve the wrong problem, with the wrong tools, the wrong way :)

- Please don't put form state in Redux
- Please don't try to write a bunch of actions that are just "blindly merge data into state"
- Please put logic _in reducers_ as much as possible

**Comment by samislam** [comment]
Hello. Sorry for this late reply,
If Redux toolkit is not recommended for keeping the form state globally,
What should I then use instead?

**Comment by markerikson** [comment]
@samislam : "form state" is almost never actually "global". There's very few times that you need controlled input data across the app.  Normally it's only needed by the one parent form component.

The better pattern is to keep the input data in component state, and dispatch _one_ action after the form is submitted / the user is done.

**Comment by samislam** [comment]
@markerikson what if you want to keep the form state when the user navigates to another page, then he returned back.
in this case it's beautiful to see the form state isn't cleared

**Comment by phryneas** [comment]
Then you safe the form away when the user navigates back, because it becomes global in that moment. Before that it is still local.
